api/buckets/service.go

```diff
@@ -596,6 +900,197 @@ func (s *Service) PushPath(server pb.API_PushPathServer) error {
+			return nil, err
+		}
+		news[i-1] = cn
+		cn = n
+	}
+	np = append(np, childNode{new: cn, name: parts[0], isLeaf: true})
+
+	// Now, we have a full list of nodes to the insert location,
+	// but the existing nodes need to be updated and re-encrypted.
+	change := make([]ipld.Node, len(np))
+	for i := len(np) - 1; i >= 0; i-- {
+		change[i] = np[i].new
+		if i > 0 {
+			p, ok := np[i-1].new.(*dag.ProtoNode)
+			if !ok {
+				return nil, dag.ErrNotProtobuf
+			}
+			if np[i].isLeaf {
+				xn, err := p.GetLinkedNode(ctx, s.IPFSClient.Dag(), np[i].name)
+				if err != nil && !errors.Is(err, dag.ErrLinkNotFound) {
+					return nil, err
+				}
+				if xn != nil {
+					np[i].old = path.IpfsPath(xn.Cid())
+					if err := p.RemoveNodeLink(np[i].name); err != nil {
+						return nil, err
+					}
+				}
+			} else {
+				xl, err := p.GetNodeLink(np[i].name)
+				if err != nil && !errors.Is(err, dag.ErrLinkNotFound) {
+					return nil, err
+				}
+				if xl != nil {
+					if err := p.RemoveNodeLink(np[i].name); err != nil {
+						return nil, err
+					}
+				}
+			}
+			if err := p.AddNodeLink(np[i].name, np[i].new); err != nil {
+				return nil, err
+			}
+			np[i-1].new, err = encryptNode(p, key)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Add all the _new_ nodes, which is the sum of the branch new ones
+	// from the missing path segment, and the changed ones from
+	// the existing path.
+	allNews := append(news, change...)
+	if err := s.IPFSClient.Dag().AddMany(ctx, allNews); err != nil {
+		return nil, err
+	}
+	// Pin brand new nodes
+	if err := s.IPFSClient.Dag().Pinning().AddMany(ctx, news); err != nil {
+		return nil, err
+	}
+	// Update changed node pins
+	for _, n := range np {
+		if n.old != nil && n.isLeaf {
+			if err := s.unpinBranch(ctx, n.old, key); err != nil {
+				return nil, err
+			}
+		}
+		if err = s.updateOrAddPin(ctx, n.old, path.IpfsPath(n.new.Cid())); err != nil {
+			return nil, err
+		}
+	}
+	return path.IpfsPath(np[0].new.Cid()), nil
+}
+
+// unpinBranch walks a the node at path, decrypting (if needed) and unpinning all nodes
+func (s *Service) unpinBranch(ctx context.Context, p path.Resolved, key []byte) error {
+	n, err := s.getNodeAtPath(ctx, p, key)
+	if err != nil {
+		return err
+	}
+	for _, l := range n.Links() {
+		if l.Name == "" {
+			continue // Data nodes will never be pinned directly
+		}
+		lp := path.IpfsPath(l.Cid)
+		if err := s.IPFSClient.Pin().Rm(ctx, lp); err != nil {
+			return err
+		}
+		if err = s.unpinBranch(ctx, lp, key); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type childNode struct {
+	old    path.Resolved
+	new    ipld.Node
+	name   string
+	isLeaf bool
+}
+
+// getNodesToPath returns a list of childNodes that point to the path,
+// The remaining path segment that was not resolvable is also returned.
+func (s *Service) getNodesToPath(ctx context.Context, base path.Resolved, pth string, key []byte) (nodes []childNode, remainder string, err error) {
```

Comment:
This is the basis for editing (adding / removing nodes) from encrypted nodes... the basic idea is to build a list of the path that leads to the needed node... it might not make it all the way, in which case the path segment is returned.